### PR TITLE
Add warning to non pickable models

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Deep copies of the model are done to make sure ORT supports models that can return different graphs in consecutive train steps

Models that are not pickable cannot be deep copied. We are adding a warning message and continue training as opposed to crash

### CODE REVIEW

File: orttraining/orttraining/python/experimental/orttrainer.py [hunk 1]
```diff
@@ -451,12 +451,19 @@ def forward(self, *inputs):
...
-            # Deepcopy inputs, since input values may change after model run.
-            sample_inputs_copy = copy.deepcopy(sample_inputs)
-            # Deepcopy model, in case model is stateful and changes after model run.
-            model_copy = copy.deepcopy(model)
+            try:
+                # Deepcopy inputs, since input values may change after model run.
+                sample_inputs_copy = copy.deepcopy(sample_inputs)
+                # Deepcopy model, in case model is stateful and changes after model run.
+                model_copy = copy.deepcopy(model)
+            except:
```
Comment (kit1980) [inline orttraining/orttraining/python/experimental/orttrainer.py:459]: Please specify the exception you want to catch, or at lease use `except Exception:`, so things like keyboard interrupt and system exit will not be caught.

**Review by kit1980 — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/python/experimental/orttrainer.py [hunk 2]
```diff
@@ -451,12 +451,19 @@ def forward(self, *inputs):
...
+                sample_inputs_copy = sample_inputs
```
Comment (BowenBao) [inline orttraining/orttraining/python/experimental/orttrainer.py:460]: Should we leave out `sample_inputs` ?

File: orttraining/orttraining/python/experimental/orttrainer.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/experimental/orttrainer.py:459]: It depends on the model. Can be TypeError, pickle.PickingError, cpickle.PicklingError, etc
Each python module can wrap the original PicklingError and raise a different exception

regardless, will change it to except Exception

**Review by kit1980 — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 3]
```diff
@@ -825,6 +825,121 @@ def testORTTrainerFrozenWeights(model_params):
...
+    torch.manual_seed(seed)
+    set_seed(seed)
+    pt_model = UnpickableModel()
+    def loss_fn(x, label):
+        return F.nll_loss(F.log_softmax(x, dim=1), label)
+    optim_config = optim.AdamConfig()
+    trainer = orttrainer.ORTTrainer(pt_model, model_desc, optim_config, loss_fn=loss_fn, options=opts)
+
+    # Train must succeed despite warning
+    _, _ = trainer.train_step(data, label)
```
Comment (spandantiwari) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:941]: Not a must-have right now, but there must be a way to test if a warning was generated. We can think about including it later.

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 4]
```diff
@@ -825,6 +825,121 @@ def testORTTrainerFrozenWeights(model_params):
     assert not all([param in session_state for param in model_params])
 
 
+@pytest.mark.parametrize("loss_scaler, optimizer_config, gradient_accumulation_steps", [
+    (None, optim.AdamConfig(), 1),
+    (None, optim.LambConfig(), 1),
+    (None, optim.SGDConfig(), 1),
+    (amp.DynamicLossScaler(), optim.AdamConfig(), 1),
+    (amp.DynamicLossScaler(), optim.LambConfig(), 5),
+    #(amp.DynamicLossScaler(), optim.SGDConfig(), 1), # SGD doesnt support fp16
```
Comment (spandantiwari) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:834]: nit: maybe remove?

**Review by spandantiwari — APPROVED** [review]
LGTM.

**Pull request merged**